modules/vfs-class-loader/src/main/java/org/apache/accumulo/classloader/vfs/context/ReloadingVFSContextClassLoaderFactory.java

```diff
@@ -309,6 +312,11 @@ protected boolean isVMInitialized() {
   @Override
   public synchronized ClassLoader getClassLoader(String contextName)
       throws IllegalArgumentException {
+    try {
+      initializeContexts();
+    } catch (URISyntaxException | IOException e) {
+      throw new RuntimeException("Error initializing contexts", e);
+    }
```

Comment:
Can you just do this in the constructor?

Reply:
Thanks for the note about the issues / PR, that's a good point that I didn't think about.

I could put this in the constructor, but if there is a failure there then I think you would have to stop the process, fix the issue, and restart the process. By putting it in getClassLoader() this allows the user to potentially fix the issue (missing file, incorrect JSON syntax, etc) with the process still up.

Reply:
I suppose it would be fixable in theory that way, but it's not likely to be handled by the containing application if it's throwing a generic RuntimeException, rather than something more specific, so it's almost certainly the case that the containing application would die anyways.

Reply:
As an example, NextBatchTask.run() catches Throwable and returns the the error to the user, it doesn't appear that it will kill the application if it can't load the iterator class.

Reply:
I don't know that's a good example, because catching Throwable isn't great practice.

I'm okay with the idea, but in general, most applications aren't going to be trying to catch RuntimeExceptions, especially generic RTEs that don't have a more specific name they know how to sanely handle. For example, an application may know how to handle IllegalArgumentException after user calling a method with user input, because it can be handled by re-prompting the user for better input. But, if it merely throws generic RuntimeException, there's not much it can know how to do, other than ignore it and log it as in the example you gave.

I just think if we want to keep things simple, it would be better to just deal with the error up front in the constructor. And, if we want something more robust and more complex, we should create a (or use an existing) specific subclass of RuntimeException that represents the specific exception that the application can watch for and handle specifically. That would set user applications up for success, being able to be written to programatically handle well-known failures, if they want to catch and handle those specific ones (if they choose to).

Reply:
To be clear, my argument summary is: keep it simple and fail in the constructor, or use a more specific exception than RuntimeException, if the intent is to be able to programatically handle the failure in the application.